File: evaluation/results/scenario_3/stan/llama3.1:8b/run_01/extracted_code.R
# Load required libraries
library(cmdstanr)
library(tidyverse)

# Load the data
data <- read_csv("observations.csv")

# Convert date to datetime format
data$date <- as_datetime(data$date, format = "%Y-%m-%d")

# Calculate delays and ascertainment rates for each stream
delays <- c(cases = 0, hospitalisations = 5, deaths = 10)
ascertains <- c(cases = 1, hospitalisations = 0.8, deaths = 0.95)

# Define the Stan model
model_code <- "
data {
  int<lower=1> S; // Number of streams (3: cases, hospitalisations, deaths)
  int<lower=1> T; // Number of time points
  vector[T] date; // Date for each time point
  array[S, T] y; // Observations for each stream and time point
  
  real<lower=0> rt_smoothness;
}

parameters {
  real<lower=0> rt[T]; // Time-varying reproduction number
  real<lower=0> mu_g; // Mean generation interval
  array[S] alpha; // Stream-specific ascertainment rates
  real<lower=0> sigma_y[S]; // Observation noise for each stream
  
  vector[3] beta; // Shared coefficients for exponential family
}

model {
  // Prior on rt_smoothness (e.g., Gaussian with mean and sd)
  rt_smoothness ~ normal(1, 2);
  
  // Priors on mu_g (e.g., Uniform between 0.5 and 10 days)
  mu_g ~ uniform(0.5, 10);
  
  // Prior on alpha (e.g., Dirichlet with concentration parameter 2)
  for (s in 1:S) {
    alpha[s] ~ dirichlet(rep(1, S));
  }
  
  // Prior on sigma_y (e.g., Half-Cauchy with scale parameter 2.5)
  for (s in 1:S) {
    sigma_y[s] ~ cauchy(0, 2.5);
  }
  
  // Priors on beta (e.g., Normal with mean and sd)
  beta ~ normal(0, 2);
  
  // Likelihood
  for (t in 1:T) {
    // Calculate observed infections using the renewal equation
    real observed_infections = 0;
    for (s in 1:S) {
      real y_observed = y[s, t] * alpha[s];
      observed_infections += y_observed / sigma_y[s]^2;
    }
    
    // Use exponential family likelihood with shared coefficients beta
    target += log_sum_exp(beta[1] + rt[t], observed_infections);
  }
  
  // Smoothness penalty on Rt (e.g., squared differences between adjacent values)
  for (t in 2:T) {
    target += pow(rt[t] - rt[t-1], 2) * rt_smoothness;
  }
}
"

# Fit the model
fit <- cmdstanr::cmd_stan_model(model_code = model_code, iter = 2000, chains = 4,
                                data = list(S = ncol(delays), T = length(unique(data$date)),
                                            date = unique(data$date),
                                            y = rbind(cases = data$cases[match(unique(data$date), data$date)],
                                                      hospitalisations = data$hospitalisations[match(unique(data$date), data$date)],
                                                      deaths = data$deaths[match(unique(data$date), data$date)]),
                                            delays = as.matrix(delays),
                                            ascertains = c(ascertains, 1 - sum(ascertains)) // include a dummy stream with ascertainment rate 1
                                ))

# Extract Rt estimates and stream-specific parameters
posterior <- extract(fit)

# Calculate Rt estimates by time point (e.g., average of posterior samples)
rt_estimates <- posterior$rt[, , ] %>%
  apply(2, mean)

# Calculate stream-specific ascertainment rates (e.g., average of posterior samples)
ascertainment_rates <- posterior$alpha[, ]

# Plot the results
plot(unique(data$date), rt_estimates,
     type = "l", main = "Time-varying Reproduction Number (Rt)",
     xlab = "Date", ylab = "Rt",
     col = "blue")
legend("topright", legend = c("Rt Estimates"),
       col = "blue",
       lty = 1, pch = 19,
       bty = "n")

# Plot the stream-specific ascertainment rates
plot(unique(data$date), ascertains[, 1], type = "l",
     main = "Stream-Specific Ascertainment Rates",
     xlab = "Date", ylab = "Ascertainment Rate")
legend("topright", legend = c("Cases", "Hospitalisations", "Deaths"),
       col = c(1, 2, 3),
       lty = 1,
       bty = "n")

# Handle the initial infection seeding period
initial_period <- data$date < as_date("2020-03-01") // assume start of pandemic

# Plot Rt estimates with initial period highlighted
plot(unique(data$date), rt_estimates, type = "l",
     main = "Time-varying Reproduction Number (Rt)",
     xlab = "Date", ylab = "Rt")
abline(v = unique(data$date)[initial_period], col = "red")

# Show the summary of the model
summary(fit)

File: evaluation/results/scenario_3/turing/llama3.1:8b/run_02/extracted_code.jl
using Turing
using DataFrames
using Statistics
using Plots
include("observations.csv")  # load data from observations.csv

# Model parameters
S = 7  # number of days in generation interval
delay_cases = 1  # delay from infection to case observation (days)
delay_hospitalisations = 5  # delay from infection to hospitalisation observation (days)
delay_deaths = 10  # delay from infection to death observation (days)

# Model

@model function rt_model(obs::DataFrame, stream_delays, stream_ascertainty_rates)
    n = length(unique(obs.date))
    Rt ~ InverseGamma(2, 1)  # prior on shared reproduction number
    β_stream_cases ~ Normal(0, 5)  # prior on ascertainment rate for cases
    β_stream_hospitalisations ~ Normal(0, 5)  # prior on ascertainment rate for hospitalisations
    β_stream_deaths ~ Normal(0, 5)  # prior on ascertainment rate for deaths

    for t in 1:n
        I_t = sum([obs[i, j] * stream_ascertainty_rates[j] *
                    stream_delays[j] > delay obs.date[t]
                   for (j, i) in enumerate(findall(obs.date .== obs.date[t])))
        rt_t ~ Beta(2, 2)  # prior on Rt at time t
        I_t ~ Poisson(Rt * sum([I_t_i * exp(-rt_t / S) *
                              exp(-(i - stream_delays[j]) / S)
                   for (j, i) in enumerate(obs[1:n, "cases", t])]))
    end

    return Rt, β_stream_cases, β_stream_hospitalisations, β_stream_deaths
end

# Load data and create observation matrix
data = readcsv("observations.csv")
dates = unique(data.date)
n_dates = length(dates)

# Create streams-specific delays and ascertainment rates
delay_cases = fill(delay_cases, n_dates);
delay_hospitalisations = fill(delay_hospitalisations, n_dates);
delay_deaths = fill(delay_deaths, n_dates);

β_stream_cases = ones(n_dates)  # assume constant ascertainment rate for cases
β_stream_hospitalisations = ones(n_dates)  # assume constant ascertainment rate for hospitalisations
β_stream_deaths = ones(n_dates)  # assume constant ascertainment rate for deaths

# Define the data structure
obs = DataFrame(
    date=repeat(dates, eachcol(data), outer=true),
    cases=data[!, "cases"],
    hospitalisations=data[!, "hospitalisations"],
    deaths=data[!, "deaths"]
)

# Fit the model
model = rt_model(obs, delay_cases, delay_hospitalisations, delay_deaths)
chain = sample(model, MCMCThreads(), 1000)

# Extract and plot results

# Rt estimates for each time point
Rt_chain = chain[:Rt]
plot(Rt_chain, label="Rt")

# Stream-specific ascertainment rate estimates
stream_ascertainty_rates = chain[[:β_stream_cases, :β_stream_hospitalisations, :β_stream_deaths]]
bar(stream_ascertainty_rates, label=["cases" "hospitalisations" "deaths"])

# Summary plot
plot!(legend=:topright)
xlabel!("Time")
ylabel!("Value")

# Display results
println("Rt estimates for each time point: ", Rt_chain)
println("Stream-specific ascertainment rate estimates: ")
println(stream_ascertainty_rates)

display(plot())

